File: qt/CuteGo/EventManager.cpp
#include <QDebug>
#include "EventManager.h"
#include <QJsonObject>
#include <QJsonDocument>
#include <QJSEngine>


#include <QDebug>

EventManager::EventManager(QQmlApplicationEngine* engine, QString name, GoCallbackFunc gocallback, QObject *parent) : QObject(parent)
{
    this->_engine = engine;
    this->_name = name;
    this->_gc = gocallback;

    connect(this, EventManager::signal_callQml, this, EventManager::slot_callQml);
}

void EventManager::call(QString eventName, QJSValue value)
{
    QJSValue jsJsonString = EventManager::jsonStringifyFunc.call({value});

    _gc(this->_name.toStdString().c_str(), eventName.toStdString().c_str(), jsJsonString.toString().toUtf8().toStdString().c_str());
}

void EventManager::on(QString eventName, QJSValue subscriber)
{


    if(!_handlers[eventName]) {
        this->_handlers[eventName] = new QVector<QJSValue>();
    }

    QVector<QJSValue>* subscribers = _handlers[eventName];

    for(auto s : *subscribers) {
        if(s.equals(subscriber)) {
            qWarning() << "CuteGo QML Warning: " << "Subscriber for '" + eventName + "\' is already registered";
            return;
        }
    }

    subscribers->push_back(subscriber);

}

void EventManager::callQml(QString eventName, QString value)
{
    emit signal_callQml(eventName, value);
}

void EventManager::slot_callQml(QString eventName, QString value)
{

    QVector<QJSValue>* subscribers = _handlers[eventName];

    if(!subscribers) {
        qWarning() << "CuteGo QML Warning: " << "Can't find handler for '" << eventName << '\'';
        return;
    }

    for(auto subscriber : *subscribers) {
        if(subscriber.isCallable()) {
            QJSValue parsedJSVal = EventManager::jsonParseFunc.call({value});
            subscriber.call({parsedJSVal});
        }
        else {
            qCritical() << "CuteGo QML Warning: " << "Can't call handler for '" << eventName << "\'. Is not callable!";
        }
    }
}

File: eventManager.go
package cutego

import "fmt"

type Subscribers []func(interface{})

type EventManager struct {
	Name     string
	Handlers map[string]Subscribers
}

func NewEventManager(name string) *EventManager {
	var em *EventManager = new(EventManager)

	(*em).Name = name
	(*em).Handlers = make(map[string]Subscribers)

	newEventManager(name)

	eventManagers[name] = em

	return em
}

func (em *EventManager) On(eventName string, subscriber func(interface{})) {

	if em.Handlers[eventName] == nil {
		em.Handlers[eventName] = Subscribers{}
	}

	for _, sub := range em.Handlers[eventName] {
		if &sub == &subscriber {
			fmt.Println("CuteGo Go Warning: Subscriber for " + eventName + " is already registered.")
			return
		}
	}

	em.Handlers[eventName] = append(em.Handlers[eventName], subscriber)

}

func (em *EventManager) Call(eventName string, value interface{}) {
	callQt(em.Name, eventName, value)
}
